### PR TITLE
[issue-561] error handling

### LINKED ISSUE #561: Update error handling and display
For all errors that are related to "metadata not relevant" - e.g. not a boosting model, no date/index column, etc. - make it a different type of error and update suite text to be something that doesn't sound like an error

For example, instead of - `DeepchecksValueError: Expected model to be a type from ['regression'], but received model of type: multiclass`
maybe something like: Check relevant for models of type ['regression'], received model of type 'multiclass'.

### PR DESCRIPTION
**Pull request by yromanyshyn**
(intended to resolve #561, see also PR #242)

This issue actually brings up a more broad concern about how we handle exceptions and dataset validation.

In general
----------------

Our `Dataset` class is loosely defined and has too many possible states. We can define a dataset with features but without a label, with a label but without features, or even entirely empty, without features and label.

This "shapelessness" of the dataset class has a negative effect. Whenever and wherever we want to do something with a dataset instance, first we need to verify presence of the features or a label. For example, if we have a check that requires a dataset with features and a label, and uses some public function from another `deepchecks` module that also requires a dataset with features and a label, then we need to do validation twice, separately in the check and separately in the public function. 

We cannot rely on the fact that whoever will use that public function, will always provide a valid instance of the dataset, and we also want to raise an error with a more specific message in the check, instead of allowing the public function to handle this.

doing the same thing twice (possible even more times) is something incorrect/

I think we should impose more restrictions upon the `Dataset` type, but I am not going to speak about it here, instead, I want to speak about how we handling this currently, and how we should change it.

We have tried to solve this problem with help of a bunch of` Dataset.validate_*` methods that we use almost everywhere when we need to validate a dataset instance. Honestly, it did not facilitate not solve the problem.
1. we just replaced two lines of code (`if ... the raise ...`) with one
2. those methods raise errors with messages that are relative only to the checks

The only reason to use those methods is to unify error messages within some scope, and that is useful only within checks scope.

What I have done and what I suggest
---------------------------------------------------

**point 1**

Added `DatasetValidationError` and `ModelValidationError` error types, that should be 
used in cases when dataset instance or model instance did not meet imposed requirements.

(to some degree it is a special case of the `DeepchecksValueError`/)

**point 2**

Removed `validate_*` methods from the `Dataset` class.

**point 3**

Added private dataset validation methods to the `BaseCheck` class that are used only
to unify error messages across all checks.

**point 4**

Added next methods to the `Dataset` class:

- `def ensure_not_empty_dataset(cls, obj: t.Any, *, cast: bool = False) -> 'Dataset': ...`
- `def datasets_share_features(cls, datasets: t.List['Dataset']) -> bool: ...`
- `def datasets_share_categorical_features(cls, datasets: t.List['Dataset']) -> bool: ...`
- `def datasets_share_label(cls, datasets: t.List['Dataset']) -> bool: ...` 

**point 5**

Rephrased some error messages by following @shir22  advice

**Note that PR is in draft state.**

**TODO:**
- [x] fix tests
- [x] add tests for new `Dataset` methods
- [x] fix comments
- [x] fix codestyle
- [x] rephrase error messages by folowing @shir22 advice (where it is possible)

**To think about:**
- should we also compare label/features dtypes within `Dataset.datasets_share_label`, `Dataset.datasets_share_features`?

### CODE REVIEW

File: deepchecks/base/dataset.py [hunk 1]
```diff
@@ -781,109 +783,213 @@ def select(
...
+            AssertionError:
+                'datasets' parameter is not instance of a list;
+                'datasets' is empty or contains only one dataset;
         """
-        Dataset.validate_dataset(other)
-        if sorted(self.features) == sorted(other.features):
-            return list(self.features)
-        else:
-            raise DeepchecksValueError('Check requires datasets to share the same features')
+        assert isinstance(datasets, list)
```
Comment (matanper) [inline deepchecks/base/dataset.py:824]: not sure I like the use of assert, what happens if it fails?

**Review by matanper — COMMENTED** [review]
I agree with your points. not sure about the classmethods on base check, a bit weird. but overall good

File: deepchecks/base/dataset.py (hunk 1 above)
Comment (yromanyshyn) [reply deepchecks/base/dataset.py:824]: it is common in the python world to use assert to verify that input parameters of the function satisfy its contract (signature). Ff `datasets` is not a list then `AssertionError` will be raised which means that function is used improperly 

I will add a message to that assert to make it cleaner what happened

File: deepchecks/base/check.py [hunk 2]
```diff
@@ -332,6 +336,67 @@ def name(cls):
...
+                'inferred incorrectly as categorical features. To fix this, manually edit the '
+                'categorical features using Dataset(cat_features=<list_of_features>'
+            )
+        return datasets[0].cat_features
+
+    @classmethod
+    def _datasets_share_label(cls, datasets: List['Dataset']) -> Hashable:
+        """TODO: add coments"""
+        if Dataset.datasets_share_label(datasets) is False:
+            raise DatasetValidationError('Check requires datasets to have and to share the same label')
```
Comment (shir22) [inline deepchecks/base/check.py:366]:
```suggestion
            raise DatasetValidationError('Check requires Datasets to have and to share the same label')
```

File: deepchecks/base/check.py [hunk 3]
```diff
@@ -332,6 +336,67 @@ def name(cls):
...
+        return cast(Hashable, datasets[0].label_name)
+
+    @classmethod
+    def _dataset_has_label(cls, dataset: Dataset) -> pd.Series:
+        """TODO: add comments"""
+        if dataset.label_col is None:
+            raise DatasetValidationError('Datasets without label are irrelevant to the check')
```
Comment (shir22) [inline deepchecks/base/check.py:373]:
```suggestion
            raise DatasetValidationError('Check is irrelevant for Datasets without label')
```

File: deepchecks/base/check.py [hunk 4]
```diff
@@ -332,6 +336,67 @@ def name(cls):
...
+        return dataset.label_col
+
+    @classmethod
+    def _dataset_has_features(cls, dataset: Dataset) -> pd.DataFrame:
+        """TODO: add comments"""
+        if (
+            dataset.features_columns is None
+            or len(dataset.features_columns.columns) == 0
+        ):
+            raise DatasetValidationError('Datasets without features are irrelevant to the check')
```
Comment (shir22) [inline deepchecks/base/check.py:383]:
```suggestion
            raise DatasetValidationError('Check is irrelevant for Datasets without features')
```

File: deepchecks/base/check.py [hunk 5]
```diff
@@ -332,6 +336,67 @@ def name(cls):
...
+        return dataset.features_columns
+
+    @classmethod
+    def _dataset_has_date(cls, dataset: Dataset) -> pd.Series:
+        """TODO: add comments"""
+        if dataset.datetime_col is None:
+            raise DatasetValidationError('Datasets without datetime column are irrelevant to the check')
```
Comment (shir22) [inline deepchecks/base/check.py:390]:
```suggestion
            raise DatasetValidationError('Check is irrelevant for Datasets without datetime column ')
```

File: deepchecks/base/check.py [hunk 6]
```diff
@@ -332,6 +336,67 @@ def name(cls):
...
+        return dataset.datetime_col
+
+    @classmethod
+    def _dataset_has_index(cls, dataset: Dataset) -> pd.Series:
+        """TODO: add comments"""
+        if dataset.index_col is None:
+            raise DatasetValidationError('Datasets without index are irrelevant to the check')
```
Comment (shir22) [inline deepchecks/base/check.py:397]:
```suggestion
            raise DatasetValidationError('Check is irrelevant for Datasets without an index')
```

File: deepchecks/base/check.py [hunk 7]
```diff
@@ -332,6 +336,67 @@ def name(cls):
         """Name of class in split camel case."""
         return split_camel_case(cls.__name__)
 
+    # NOTE: next set of private functions exists to unify error messages across all checks
```
Comment (shir22) [inline deepchecks/base/check.py:339]: General note - 
One more error that is currently a "DeepchecksValueError" is in the BoostingOverfit check... 
(if model is not of type boosting. not sure where this should be but it also belongs to these types of "irrelevant data/model for check")

**Review by ItayGabbay — COMMENTED** [review]
I agree with your comments.
One important note: We currently add computer vision checks as a namespaced package. In addition, I moved all current checks and suits to be in a `tabular` namespaced packages, and some the core functionality and classes (including basecheck, basesuites, etc.) to a `base` namespaced package. (This is currently a draft).

I wonder where the validations should happen? I don't think it should be inside the base subpackage since they are specific to tabular datasets and not related to CV.

File: deepchecks/base/dataset.py (hunk 1 above)
Comment (yromanyshyn) [reply deepchecks/base/dataset.py:824]: oh, also regarding this

> ..not sure about the classmethods ..

they do not depend upon instance variables (state), so I decided to declare thems as classmethods

**Comment by yromanyshyn** [comment]
@ItayGabbay 

> wonder where the validations should happen? I don't think it should be inside the base subpackage since they are specific to tabular datasets and not related to CV.

oh/ yup you are right
then it will be better to put those methods into separate modules, like
- `deepchecks.tabular.checks._dataset_validation`
- `deepchecks.vision.checks._dataset_validation`

something like that

File: deepchecks/base/check.py [hunk 8]
```diff
@@ -332,6 +338,157 @@ def name(cls):
...
+
+        Raises:
+            DatasetValidationError: if dataset does not have index;
+        """
+        if dataset.index_col is None:
+            raise DatasetValidationError('Check is irrelevant for Datasets without an index')
+        return dataset.index_col
+
+    @staticmethod
+    def _verify_model_type(
```
Comment (JKL98ISR) [inline deepchecks/base/check.py:471]: why it should be in the Check class? I think will be better in utild or dataset class

Comment (yromanyshyn) [reply deepchecks/base/check.py:471]: - it is used only within checks to unify error message
- it raises  an exception with check related message

**Comment by ItayGabbay** [comment]
image

File: deepchecks/base/check.py [hunk 9]
```diff
@@ -9,30 +9,36 @@
 # ----------------------------------------------------------------------------
 #
 """Module containing all the base classes for checks."""
-# pylint: disable=broad-except
+# pylint: disable=broad-except,inconsistent-quotes
```
Comment (benisraeldan) [inline deepchecks/base/check.py:12]: why disable this lint?

**Comment by yromanyshyn** [comment]
@ItayGabbay did not get you? was not it expected? (error name is not show only for `DatasetValidationError`, `ModelValidationError` and `DeepchecksProcessError`)

**Comment by ItayGabbay** [comment]
> @ItayGabbay did not get you? was not it expected? (error name is not show only for `DatasetValidationError`, `ModelValidationError` and `DeepchecksProcessError`)

It's exactly like how I expected. I attached the screenshot so everyone will see the output :) 
great work.

File: deepchecks/checks/methodology/identifier_leakage.py [hunk 10]
```diff
@@ -51,14 +52,17 @@ def run(self, dataset: Dataset, model=None) -> CheckResult:
...
+        self._dataset_has_label(dataset)
 
+        ppscore_params = ppscore_params or {}
         relevant_columns = list(filter(None, [dataset.datetime_name, dataset.index_name, dataset.label_name]))
 
         if len(relevant_columns) == 1:
-            raise DeepchecksValueError('Dataset needs to have a date or index column.')
+            raise DatasetValidationError(
+                'Datasets without index or date column '
+                'are irrelevant to the current check'
```
Comment (shir22) [inline deepchecks/checks/methodology/identifier_leakage.py:64]:
```suggestion
                'Check is irrelevant for '
                'Datasets without index or date column'
```

File: deepchecks/checks/distribution/trust_score_comparison.py [hunk 11]
```diff
@@ -72,35 +72,40 @@ def run(self, train_dataset, test_dataset, model=None) -> CheckResult:
...
-        train_dataset.validate_shared_features(test_dataset)
+        train_dataset = Dataset.ensure_not_empty_dataset(train_dataset)
+        train_label = self._dataset_has_label(train_dataset)
+
+        features_list = self._datasets_share_features([train_dataset, test_dataset])
+        label_name = self._datasets_share_label([train_dataset, test_dataset])
 
         if test_dataset.n_samples < self.min_test_samples:
             msg = ('Number of samples in test dataset have not passed the minimum. you can change '
                    'minimum samples needed to run with parameter "min_test_samples"')
```
Comment (shir22) [inline deepchecks/checks/distribution/trust_score_comparison.py:89]:
```suggestion
            msg = ('Number of samples in test dataset has not passed the minimum. You can change '
                   'the minimum number of samples required for the check to run with the parameter "min_test_samples"')
```

File: deepchecks/base/check.py (hunk 9 above)
Comment (yromanyshyn) [reply deepchecks/base/check.py]: I wanted to disable it for two specific lines, but when I added directive at the end of the lines it caused `line too long` error.
Yup, you are right) it is better to use it only for that lines, fixed

File: deepchecks/checks/methodology/boosting_overfit.py [hunk 12]
```diff
@@ -61,7 +61,7 @@ def predict_proba(self, x):
         elif self.model_class == 'CatBoostClassifier':
             return self.model.predict_proba(x, ntree_end=self.step)
         else:
-            raise DeepchecksValueError(f'Unsupported model of type: {self.model_class}')
+            raise ModelValidationError(f'Unsupported model of type: {self.model_class}')
```
Comment (shir22) [inline deepchecks/checks/methodology/boosting_overfit.py:64]: This error message (and also in lines 77 and 93) should be
"Check is relevant for Boosting models of type <list_of_supported model types>, but received model of type <model_class>"

File: tests/checks/performance/calibration_score_test.py [hunk 13]
```diff
@@ -11,29 +11,39 @@
...
     # Assert
     train, _, clf = diabetes_split_dataset_and_model
-    assert_that(calling(CalibrationScore().run).with_args(train, clf),
-                raises(DeepchecksValueError, r'Expected model to be a type from'
-                                           r' \[\'multiclass\', \'binary\'\], but received model of type: regression'))
+    assert_that(
+        calling(CalibrationScore().run).with_args(train, clf),
+        raises(
+            ModelValidationError,
+            r'Check relevant for models of type \[\'multiclass\', \'binary\'\], '
```
Comment (shir22) [inline tests/checks/performance/calibration_score_test.py:43]:
```suggestion
            r'Check is relevant for models of type \[\'multiclass\', \'binary\'\], '
```

File: deepchecks/base/check.py [hunk 14]
```diff
@@ -332,6 +338,157 @@ def name(cls):
...
+        Returns:
+            ModelType: type of the provided model
+
+        Raises:
+            ModelValidationError: if unexpected model type is provided;
+        """
+        task_type = task_type_check(model, dataset)
+        if task_type not in expected_types:
+            raise ModelValidationError(
+                f'Check relevant for models of type {[e.value.lower() for e in expected_types]}, '
```
Comment (shir22) [inline deepchecks/base/check.py:487]:
```suggestion
                f'Check is relevant for models of type {[e.value.lower() for e in expected_types]}, '
```

File: tests/checks/performance/regression_error_distribution_test.py [hunk 15]
```diff
@@ -14,30 +14,37 @@
...
     # Assert
     _, test, clf = iris_split_dataset_and_model
-    assert_that(calling(RegressionErrorDistribution().run).with_args(test, clf),
-                raises(DeepchecksValueError, r'Expected model to be a type from'
-                                           r' \[\'regression\'\], but received model of type: multiclass'))
+    assert_that(
+        calling(RegressionErrorDistribution().run).with_args(test, clf),
+        raises(
+            ModelValidationError, 
+            r'Check relevant for models of type \[\'regression\'\], '
```
Comment (shir22) [inline tests/checks/performance/regression_error_distribution_test.py:45]:
```suggestion
            r'Check is relevant for models of type \[\'regression\'\], '
```

File: tests/checks/performance/roc_report_test.py [hunk 16]
```diff
@@ -17,30 +17,37 @@
...
     # Assert
     train, _, clf = diabetes_split_dataset_and_model
-    assert_that(calling(RocReport().run).with_args(train, clf),
-                raises(DeepchecksValueError, r'Expected model to be a type from'
-                                           r' \[\'multiclass\', \'binary\'\], but received model of type: regression'))
+    assert_that(
+        calling(RocReport().run).with_args(train, clf),
+        raises(
+            ModelValidationError, 
+            r'Check relevant for models of type '
```
Comment (shir22) [inline tests/checks/performance/roc_report_test.py:48]:
```suggestion
            r'Check is relevant for models of type '
```

**Review by shir22 — CHANGES_REQUESTED** [review]

**Comment by shir22** [comment]
and +1 to Itay's comment, completely agree
the print screen did enable to see it in action (and therefore my minor phrasing comments), anyway looks great!

File: deepchecks/checks/methodology/identifier_leakage.py (hunk 10 above)
Comment (yromanyshyn) [reply deepchecks/checks/methodology/identifier_leakage.py]: fixed

File: deepchecks/checks/distribution/trust_score_comparison.py (hunk 11 above)
Comment (yromanyshyn) [reply deepchecks/checks/distribution/trust_score_comparison.py]: fixed

File: deepchecks/checks/methodology/boosting_overfit.py (hunk 12 above)
Comment (yromanyshyn) [reply deepchecks/checks/methodology/boosting_overfit.py]: fixed

File: tests/checks/performance/calibration_score_test.py (hunk 13 above)
Comment (yromanyshyn) [reply tests/checks/performance/calibration_score_test.py]: fixed

File: deepchecks/base/check.py (hunk 14 above)
Comment (yromanyshyn) [reply deepchecks/base/check.py]: fixed

File: tests/checks/performance/confusion_matrix_report_test.py [hunk 17]
```diff
@@ -12,29 +12,38 @@
...
     # Assert
     train, _, clf = diabetes_split_dataset_and_model
-    assert_that(calling(ConfusionMatrixReport().run).with_args(train, clf),
-                raises(DeepchecksValueError, r'Expected model to be a type from'
-                                           r' \[\'multiclass\', \'binary\'\], but received model of type: regression'))
+    assert_that(
+        calling(ConfusionMatrixReport().run).with_args(train, clf),
+        raises(
+            ModelValidationError,
+            r'Check relevant for models of type \[\'multiclass\', \'binary\'\], '
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply tests/checks/performance/confusion_matrix_report_test.py]: fixed

File: tests/checks/performance/regression_systematic_error_test.py [hunk 18]
```diff
@@ -9,38 +9,41 @@
...
     # Assert
     _, test, clf = iris_split_dataset_and_model
-    assert_that(calling(RegressionSystematicError().run).with_args(test, clf),
-                raises(DeepchecksValueError, r'Expected model to be a type from'
-                                             r' \[\'regression\'\], but received model of type: multiclass'))
+    assert_that(
+        calling(RegressionSystematicError().run).with_args(test, clf),
+        raises(
+            ModelValidationError, 
+            r'Check relevant for models of type '
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply tests/checks/performance/regression_systematic_error_test.py]: fixed

File: tests/checks/performance/roc_report_test.py (hunk 16 above)
Comment (yromanyshyn) [reply tests/checks/performance/roc_report_test.py]: fixed

**Comment by yromanyshyn** [comment]
@shir22  I fixed your remarks, thanks!

**Review by shir22 — APPROVED** [review]

**Review by ItayGabbay — APPROVED** [review]

**Pull request merged**